From b965ea5d23415d161470ac03459af34715918424 Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Wed, 20 Apr 2016 15:40:43 +0800
Subject: [PATCH 708/752] mmc: sd: fix staying busy issue for some cards

Some SD cards need delay before ACMD41 to avoid failed power up.
So move the delay to the front of the for loop which repeatly
issues the ACMD41.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
[Original patch taken from QorIQ-SDK-V2.0-20160527-yocto]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/mmc/core/sd_ops.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/core/sd_ops.c b/drivers/mmc/core/sd_ops.c
index 16b774c..1558028 100644
--- a/drivers/mmc/core/sd_ops.c
+++ b/drivers/mmc/core/sd_ops.c
@@ -157,6 +157,13 @@ int mmc_send_app_op_cond(struct mmc_host *host, u32 ocr, u32 *rocr)
 	cmd.flags = MMC_RSP_SPI_R1 | MMC_RSP_R3 | MMC_CMD_BCR;
 
 	for (i = 100; i; i--) {
+		/*
+		 * Some cards need delay before ACMD41 to avoid
+		 * failed power up. So move the delay to the front
+		 * of the for loop.
+		 */
+		mmc_delay(10);
+
 		err = mmc_wait_for_app_cmd(host, NULL, &cmd, MMC_CMD_RETRIES);
 		if (err)
 			break;
@@ -175,8 +182,6 @@ int mmc_send_app_op_cond(struct mmc_host *host, u32 ocr, u32 *rocr)
 		}
 
 		err = -ETIMEDOUT;
-
-		mmc_delay(10);
 	}
 
 	if (!i)
-- 
2.9.3

